things/service/src/main/java/org/eclipse/ditto/things/service/persistence/actors/strategies/commands/MigrateThingDefinitionStrategy.java

```diff
@@ -0,0 +1,279 @@
+                                    metadata
+                            );
+
+                            return CompletableFuture.completedFuture(mergeResult);
+                        })
+        );
+    }
+
+    private JsonObject evaluatePatchConditions(final Thing existingThing,
+            final JsonObject migrationPayload,
+            final Map<ResourceKey, String> patchConditions,
+            final DittoHeaders dittoHeaders) {
+        final JsonObjectBuilder adjustedPayloadBuilder = migrationPayload.toBuilder();
+
+        for (Map.Entry<ResourceKey, String> entry : patchConditions.entrySet()) {
+            final ResourceKey resourceKey = entry.getKey();
+            final String conditionExpression = entry.getValue();
+
+            final boolean conditionMatches = evaluateCondition(existingThing, conditionExpression, dittoHeaders);
+
+            final JsonPointer resourcePointer = JsonFactory.newPointer(resourceKey.getResourcePath());
+            if (!conditionMatches && doesMigrationPayloadContainResourceKey(migrationPayload, resourcePointer)) {
+                adjustedPayloadBuilder.remove(resourcePointer);
+            }
+        }
+
+        return adjustedPayloadBuilder.build();
+    }
+    public boolean doesMigrationPayloadContainResourceKey(JsonObject migrationPayload, JsonPointer pointer) {
+        return migrationPayload.getValue(pointer).isPresent();
+    }
+    private boolean evaluateCondition(final Thing existingThing,
+            final String conditionExpression,
+            final DittoHeaders dittoHeaders) {
+        try {
+            final var criteria = QueryFilterCriteriaFactory
+                    .modelBased(RqlPredicateParser.getInstance())
+                    .filterCriteria(conditionExpression, dittoHeaders);
+
+            final var predicate = ThingPredicateVisitor.apply(criteria,
+                    PlaceholderFactory.newPlaceholderResolver(TIME_PLACEHOLDER, new Object()));
+
+            return predicate.test(existingThing);
+        } catch (Exception e) {
+            return false;
+        }
+    }
+
+    private CompletionStage<Thing> generateSkeleton(final Context<ThingId> context,
+            final MigrateThingDefinition command,
+            final DittoHeaders dittoHeaders) {
+        return wotThingSkeletonGenerator.provideThingSkeletonForCreation(
+                        command.getEntityId(),
+                        ThingsModelFactory.newDefinition(command.getThingDefinitionUrl()),
+                        dittoHeaders
+                )
+                .thenApply(optionalSkeleton -> optionalSkeleton.orElseThrow());
+    }
+
+
+    private Thing extractDefinitions(final Thing thing, final String thingDefinitionUrl) {
+        var thingBuilder = ThingsModelFactory.newThingBuilder();
+        thingBuilder.setDefinition(ThingsModelFactory.newDefinition(thingDefinitionUrl));
+        thing.getFeatures().orElseGet(null).forEach(feature-> {
```

Comment:
you cannot pass `null` in `orElseGet` - this will lead to a NPE, as a (non-nullable) `Supplier` is needed